File: protocol/go/renderer.go
package lse

import (
	"fmt"
	"sort"
	"strings"
)

// RenderExplicit serializes a SemanticNode as explicit bracket syntax.
func RenderExplicit(node *SemanticNode) string {
	// Handle compound nodes
	if node.Kind == KindCompound {
		parts := make([]string, len(node.Statements))
		for i, stmt := range node.Statements {
			parts[i] = RenderExplicit(&stmt)
		}
		chain := node.ChainType
		if chain == "" {
			chain = "then"
		}
		return strings.Join(parts, " "+chain+" ")
	}

	parts := []string{node.Action}

	// Sort roles alphabetically for deterministic output
	roleNames := make([]string, 0, len(node.Roles))
	for role := range node.Roles {
		roleNames = append(roleNames, role)
	}
	sort.Strings(roleNames)

	// Add roles
	for _, role := range roleNames {
		value := node.Roles[role]
		if value.Type == TypeFlag {
			prefix := "+"
			if !value.Enabled {
				prefix = "~"
			}
			parts = append(parts, prefix+value.Name)
		} else {
			parts = append(parts, role+":"+valueToString(value))
		}
	}

	// Handle event handler body
	if node.Kind == KindEventHandler && len(node.Body) > 0 {
		bodyParts := make([]string, len(node.Body))
		for i, b := range node.Body {
			bodyParts[i] = RenderExplicit(&b)
		}
		parts = append(parts, "body:"+strings.Join(bodyParts, " "))
	}

	// Conditional branches (v1.1)
	if len(node.ThenBranch) > 0 {
		branchParts := make([]string, len(node.ThenBranch))
		for i, b := range node.ThenBranch {
			branchParts[i] = RenderExplicit(&b)
		}
		parts = append(parts, "then:"+strings.Join(branchParts, " "))
	}
	if len(node.ElseBranch) > 0 {
		branchParts := make([]string, len(node.ElseBranch))
		for i, b := range node.ElseBranch {
			branchParts[i] = RenderExplicit(&b)
		}
		parts = append(parts, "else:"+strings.Join(branchParts, " "))
	}

	// Loop fields (v1.1)
	if node.LoopVariant != "" {
		parts = append(parts, "loopVariant:"+node.LoopVariant)
	}
	if len(node.LoopBody) > 0 {
		bodyParts := make([]string, len(node.LoopBody))
		for i, b := range node.LoopBody {
			bodyParts[i] = RenderExplicit(&b)
		}
		parts = append(parts, "loop-body:"+strings.Join(bodyParts, " "))
	}
	if node.LoopVariable != "" {
		parts = append(parts, "loopVariable:"+fmt.Sprintf("%q", node.LoopVariable))
	}
	if node.IndexVariable != "" {
		parts = append(parts, "indexVariable:"+fmt.Sprintf("%q", node.IndexVariable))
	}

	return "[" + strings.Join(parts, " ") + "]"
}

// needsSelectorWrap reports whether a selector must be wrapped in "<.../>" to
// survive a re-parse.
//
// A bare token re-classifies as a selector only if it starts with one of
// '# . @ *'. A space would split the token, and a leading '[' in a structural role
// would re-parse as a nested command -- so both must wrap. ".a>.b" needs no
// wrapping: it has no space and keeps its leading '.'.
func needsSelectorWrap(value string) bool {
	if strings.Contains(value, " ") {
		return true
	}
	switch value[0] {
	case '#', '.', '@', '*':
		return false
	}
	return true
}

// valueToString converts a SemanticValue to its explicit syntax string form.
func valueToString(v SemanticValue) string {
	switch v.Type {
	case TypeLiteral:
		if v.Value == nil {
			return ""
		}
		switch val := v.Value.(type) {
		case string:
			if v.DataType == "string" || strings.ContainsAny(val, " \t\n\r") {
				return fmt.Sprintf("%q", val)
			}
			return val
		case bool:
			if val {
				return "true"
			}
			return "false"
		case int:
			return fmt.Sprintf("%d", val)
		case int64:
			return fmt.Sprintf("%d", val)
		case float64:
			if val == float64(int64(val)) {
				return fmt.Sprintf("%d", int64(val))
			}
			return fmt.Sprintf("%g", val)
		default:
			return fmt.Sprintf("%v", val)
		}

	case TypeSelector:
		if v.Value == nil {
			return ""
		}
		s := fmt.Sprintf("%v", v.Value)
		if s == "" {
			return s
		}
		if needsSelectorWrap(s) {
			return "<" + s + "/>"
		}
		return s

	case TypeReference:
		if v.Value == nil {
			return ""
		}
		return fmt.Sprintf("%v", v.Value)

	case TypeExpression, TypePropertyPath:
		return v.Raw

	case TypeFlag:
		return v.Name

	default:
		if v.Value != nil {
			return fmt.Sprintf("%v", v.Value)
		}
		return ""
	}
}

File: protocol/typescript/src/renderer.ts
import { type SemanticNode, type SemanticValue } from './types';

/** Serializes a SemanticNode as explicit bracket syntax. */
export function renderExplicit(node: SemanticNode): string {
  if (node.kind === 'compound') {
    const parts = (node.statements ?? []).map(renderExplicit);
    const chain = node.chainType ?? 'then';
    return parts.join(` ${chain} `);
  }

  const parts: string[] = [node.action];

  // Sort roles alphabetically for deterministic output
  const roleNames = Object.keys(node.roles).sort();

  for (const role of roleNames) {
    const value = node.roles[role];
    if (value.type === 'flag') {
      const prefix = value.enabled !== false ? '+' : '~';
      parts.push(prefix + value.name);
    } else {
      parts.push(`${role}:${valueToString(value)}`);
    }
  }

  // Event handler body (appended after roles)
  if (node.kind === 'event-handler' && node.body && node.body.length > 0) {
    const bodyStr = node.body.map(renderExplicit).join(' ');
    parts.push(`body:${bodyStr}`);
  }

  // Conditional branches (v1.1)
  if (node.thenBranch && node.thenBranch.length > 0) {
    parts.push(`then:${node.thenBranch.map(renderExplicit).join(' ')}`);
  }
  if (node.elseBranch && node.elseBranch.length > 0) {
    parts.push(`else:${node.elseBranch.map(renderExplicit).join(' ')}`);
  }

  // Loop fields (v1.1)
  if (node.loopVariant) {
    parts.push(`loopVariant:${node.loopVariant}`);
  }
  if (node.loopBody && node.loopBody.length > 0) {
    parts.push(`loop-body:${node.loopBody.map(renderExplicit).join(' ')}`);
  }
  if (node.loopVariable) {
    parts.push(`loopVariable:${JSON.stringify(node.loopVariable)}`);
  }
  if (node.indexVariable) {
    parts.push(`indexVariable:${JSON.stringify(node.indexVariable)}`);
  }

  return `[${parts.join(' ')}]`;
}

/** Converts a SemanticValue to its explicit syntax string form. */
export function valueToString(v: SemanticValue): string {
  switch (v.type) {
    case 'literal': {
      if (v.value === undefined || v.value === null) return '';
      if (typeof v.value === 'boolean') return v.value ? 'true' : 'false';
      if (typeof v.value === 'number') return String(v.value);
      if (typeof v.value === 'string') {
        // String literals are always quoted; other dataTypes (duration) are not
        if (v.dataType === 'string') return JSON.stringify(v.value);
        return v.value;
      }
      return String(v.value);
    }

    case 'selector':
    case 'reference':
      return typeof v.value === 'string' ? v.value : String(v.value ?? '');

    case 'expression':
    case 'property-path':
      return v.raw ?? '';

    case 'flag':
      return v.name ?? '';

    default:
      return v.value !== undefined ? String(v.value) : '';
  }
}
